### PR TITLE
cage: 0.1.2.1 -> 0.1.3

### PR DESCRIPTION
**Pull request by Synthetica9**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Synthetica9** [comment]
On test: 

```
machine: waiting for alice@machine to appear on screen
machine: performing optical character recognition
machine: sending monitor command: screendump /build/tmp5er4se9f
machine # [   14.433187] show_signal_msg: 93 callbacks suppressed
machine # [   14.433191] .cage-wrapped[753]: segfault at 20 ip 00007f0325aa78c7 sp 00007ffc7af59790 error 4 in libwlroots.so.8[7f0325aa1000+5a000]
machine # [   14.436361] Code: 4f a8 ff ff eb aa 0f 1f 44 00 00 c3 0f 1f 80 00 00 00 00 41 54 49 89 f4 55 31 ed 53 48 89 fb 48 8d 7f 18 48 8d 83 b8 00 00 00 <80> 7f 08 00 75 0d 48 83 3f 00 0f 85 91 00 00 00 48 89 fd 48 83 c7
machine # [   14.360139] systemd[1]: Created slice system-systemd\x2dcoredump.slice.
machine # [   14.366883] systemd[1]: Started Process Core Dump (PID 888/UID 0).
machine # [   14.654872] systemd-coredump[889]: Process 753 (.cage-wrapped) of user 1000 dumped core.
```

Oof!

**Comment by primeos** [comment]
The CI test also fails (on both architectures) so it seems reproducible and doesn't only affect your system. Have you tried reproducing the test by running `xterm` manually or via the Cage module? (Unfortunately it could even depend on the kernel/mesa driver.)

**Comment by Synthetica9** [comment]
> The CI test also fails (on both architectures) so it seems reproducible and doesn't only affect your system. Have you tried reproducing the test by running `xterm` manually or via the Cage module? (Unfortunately it could even depend on the kernel/mesa driver.)

I have run `xterm` manually, not yet via the cage module. I'll try that I guess

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 119615` at b1a2ebe7 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>cage</li>
    <li>wio</li>
  </ul>
</details>

**Comment by yorickvP** [comment]
I'm also seeing this segfault in the test. It would be useful to have a backtrace.

**Comment by primeos** [comment]
I was about to take a look at the test but it is already broken since 2021-02-20 (no segfault though): https://hydra.nixos.org/build/141768308

cc @matthewbauer @flokli (test maintainers): Can you take a look?

**Comment by flokli** [comment]
I'll start bisecting this.

**Comment by berbiche** [comment]
There's also a stacktrace available here: https://github.com/NixOS/nixpkgs/pull/119963#issuecomment-823439106

**Comment by flokli** [comment]
I gave up trying to bisect this - there was some other issues which made this pretty much impossible to bisect.

It might be related to the `WAYLAND_DISPLAY` changes - see https://github.com/NixOS/nixpkgs/issues/119445#issuecomment-819766445 for context.

**Comment by flokli** [comment]
Note https://github.com/NixOS/nixpkgs/pull/121174 - I can't/want to really debug/maintain this without real hardware.

**Comment by primeos** [comment]
> I gave up trying to bisect this - there was some other issues which made this pretty much impossible to bisect.

Ok but thanks for the attempt :)

> It might be related to the WAYLAND_DISPLAY changes - see #119445 (comment) for context.

I also wondered about that but that change was only in Sway and the test checks for `/run/user/${toString user.uid}/wayland-0.lock` so it should be fine.

I've tried out a few things (`programs.xwayland.enable=true`, using Wayland terminals like weston-terminal and foot, and setting `Restart=on-failure` for the systemd services but Cage always segfaults).

However when I use the Cage module on my test PC everything works. So either I'm having something else enabled that's required or it only fails inside the VM, e.g. due to different GPU drivers.

**Comment by primeos** [comment]
@GrahamcOfBorg test cage

The test should finally succeed now... Seems like it's indeed related to the GPU driver. I guess switching to `virtio` should be fine (is faster anyway). Needed to rebase as I've fixed the existing test failure in #121190.

**Comment by Synthetica9** [comment]
Does this also fix the cagebreak module?

**Review by Synthetica9 — COMMENTED** [review]
Can confirm that tests work again.

**Review by primeos — APPROVED** [review]
Yay, should be finally good to go then :)

> Does this also fix the cagebreak test?

Didn't test it but yes, it should (at least if it's the same / a similar segfault). Feel free to reuse the fix.

Seems like wlroots 0.13 had some change that breaks with `-vga std` (the default). But that fully virtualized driver is so slow that likely almost no one uses that configuration anyway (and it might also depend on the Mesa version, kernel drivers, etc.).

**Comment by Synthetica9** [comment]
> Seems like wlroots 0.13 had some change that breaks with -vga std (the default).

Is that something we want to report upstream?

File: nixos/tests/cage.nix [hunk 1]
```diff
@@ -25,6 +25,11 @@ import ./make-test-python.nix ({ pkgs, ...} :
   testScript = { nodes, ... }: let
     user = nodes.machine.config.users.users.alice;
   in ''
+    # Need to switch to a different VGA card / GPU driver because Cage segfaults with the default one (std):
+    # machine # [   14.355893] .cage-wrapped[736]: segfault at 20 ip 00007f035fa0d8c7 sp 00007ffce9e4a2f0 error 4 in libwlroots.so.8[7f035fa07000+5a000]
+    # machine # [   14.358108] Code: 4f a8 ff ff eb aa 0f 1f 44 00 00 c3 0f 1f 80 00 00 00 00 41 54 49 89 f4 55 31 ed 53 48 89 fb 48 8d 7f 18 48 8d 83 b8 00 00 00 <80> 7f 08 00 75 0d 48 83 3f 00 0f 85 91 00 00 00 48 89 fd 48 83 c7
+    os.environ["QEMU_OPTS"] = "-vga virtio"
```
[the remark this answers is not in the document]
Comment (primeos) [reply nixos/tests/cage.nix:31]: Not sure what you mean here. It's definitely a hack but from a very brief look I haven't found another easy way to pass this though to the test driver script.

**Comment by primeos** [comment]
> Is that something we want to report upstream?

I also wondered about that but I think we should at least provide a full stack trace and if this only affects that one driver/combination is likely pretty much irrelevant. But could maybe be something to mention on their IRC? Anyway, feel free to contact them.

File: nixos/tests/cage.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (flokli) [reply nixos/tests/cage.nix:31]: `QEMU_OPTS` is something that should be set by people running VM tests interactively, to do port forwardings etc.

We really should not poke with it from the test script - that's what `virtualisation.qemu.options` is for.

@Synthetica9 @primeos can one of you send another PR to clean this up?

[the remark this answers is not in the document]
Comment (primeos) [reply nixos/tests/cage.nix:31]: Opened #121416, thanks :)

I figured that `QEMU_OPTS` isn't really intended for such usages but I wasn't aware of  `virtualisation.qemu.options` (both of them are apparently completely undocumented, at least in Nixpkgs, which seems to be the main problem).

**Pull request merged**